### PR TITLE
Revision negotiation

### PR DESCRIPTION
**Pull request by pmelab**

### CODE REVIEW

File: tests/src/Kernel/RevisionNegotiationTest.php [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+      ->setRevisionable(TRUE)
+      ->setProvider('entity_test_rev');
+
+    $c = BaseFieldDefinition::create('string')
+      ->setName('c')
+      ->setRevisionable(TRUE)
+      ->setProvider('entity_test_rev');
+
+    $d = BaseFieldDefinition::create('string')
+      ->setName('c')
```
Comment (plach79) [inline tests/src/Kernel/RevisionNegotiationTest.php]:
```suggestion
      ->setName('d')
```

File: tests/src/Kernel/RevisionNegotiationTest.php [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+   * A negative context weight triggers an inverse match.
+   * Non-matching revisions are ranked down instead of matching revisions up.
+   */
+  public function testNegativeMatching() {
+    /** @var \Drupal\revision_tree\Entity\EntityRepository $repository */
+    $repository = $this->container->get('entity.repository');
+    /** @var \Drupal\Core\Entity\ContentEntityStorageInterface $storage */
+    $storage = $this->entityManager->getStorage('entity_test_rev');
+
+    /** @var \Drupal\Core\Entity\ContentEntityInterface $a */
```
Comment (plach79) [inline tests/src/Kernel/RevisionNegotiationTest.php]:
```suggestion
    /** @var \Drupal\Core\Entity\ContentEntityInterface $x */
```

File: src/Entity/EntityRepository.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+class EntityRepository extends CoreEntityRepository {
+
+  public function __construct(
+    EntityTypeManagerInterface $entity_type_manager,
+    LanguageManagerInterface $language_manager
+  ) {
+    parent::__construct($entity_type_manager, $language_manager);
+  }
+
+  public function getActive(EntityInterface $entity, array $contexts) {
```
Comment (plach79) [inline src/Entity/EntityRepository.php]: `$context` should be an array of `\Drupal\Core\Plugin\Context\ContextInterface`.

File: src/EntityQuery/Query.php [hunk 4]
```diff
@@ -0,0 +1,160 @@
+<?php
+
+namespace Drupal\revision_tree\EntityQuery;
```
Comment (plach79) [inline src/EntityQuery/Query.php]: Ideally this would be `Drupal\revision_tree\Entity\Query\Sql` :)

File: src/EntityQuery/Query.php [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  /**
+   * Order the results by matching context fields.
+   *
+   * @param array $contexts
+   *   An array of context values keyed by context id.
+   * @param string $direction
+   *   Either 'ASC' or 'DESC'.
+   * @param string | null $langcode
+   */
+  public function orderByContextMatching(array $contexts, $direction = 'DESC', $langcode = NULL) {
```
Comment (plach79) [inline src/EntityQuery/Query.php]: What about merging `$langcode` into `$contexts`?

**Comment by plach79** [comment]
While running tests and I got the following error (on MySql 5.5):
```
Drupal\Core\Database\DatabaseExceptionWrapper : SQLSTATE[42000]: Syntax error or access violation: 1055 'drupal_8x_dev.base_table.revision_id' isn't in GROUP BY: SELECT base_table.revision_id AS revision_id, base_table.id AS id, 0 + IF(entity_test_rev_revision.a = :a__value, 1, 0) * :a__weight AS revision_context_match, CONCAT('[', GROUP_CONCAT(CONCAT('["', REPLACE(base_table.revision_id, '"', '\"'), '",', 0 + IF(entity_test_rev_revision.a = :a__value, 1, 0) * :a__weight, ']')), ']') AS revision_id
FROM
{entity_test_rev_revision} base_table
LEFT JOIN {entity_test_rev_revision} entity_test_rev_revision ON entity_test_rev_revision.revision_id = base_table.revision_id
INNER JOIN {entity_test_rev} entity_test_rev ON entity_test_rev.id = base_table.id
WHERE entity_test_rev.id = :db_condition_placeholder_0
GROUP BY base_table.id
ORDER BY revision_context_match DESC; Array
(
    [:db_condition_placeholder_0] => 1
    [:a__value] => x
    [:a__weight] => 1
)
```

See https://dev.mysql.com/doc/refman/5.5/en/sql-mode.html#sqlmode_only_full_group_by. The `ONLY_FULL_GROUP_BY` is enabled by default by the MySql driver.

**Comment by amateescu** [comment]
Thanks @pmelab for this effort :)

**Pull request merged**